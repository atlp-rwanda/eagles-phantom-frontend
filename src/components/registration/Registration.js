import React, { Component } from "react";
import { connect } from "react-redux";
import {
  Form,
  Col,
  Button,
  Badge,
  Spinner,
  InputGroup,
  FormControl,
} from "react-bootstrap/";
import NavigationBar from "../../components/layouts/NavigationBar";
import Footer from '../layouts/Footer';
import { registrationAction } from "../../redux/actions/registrationAction";

class Registration extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: "",
      firstname: "",
      lastname: "",
      gender: "",
      address: "",
      role: "",
      dateofbirth: "",
      emailError: "",
      firstnameError: "",
      lastnameError: "",
      genderError: "",
      addressError: "",
      roleError: "",
      dateofbirthError: "",
      emailErrorStatus: false,
      firstnameErrorStatus: false,
      lastnameErrorStatus: false,
      genderErrorStatus: false,
      addressErrorStatus: false,
      roleErrorStatus: false,
      dateofbirthErrorStatus: false,
    };
  }

  handleSubmit = () => {
    this.setState(() => this.initialState);
  };

  validation = () => {
    let isError = false;

    const errors = {
      emailError: "",
      firstnameError: "",
      lastnameError: "",
      genderError: "",
      addressError: "",
      roleError: "",
      dateofbirthError: "",
      emailErrorStatus: false,
      firstnameErrorStatus: false,
      lastnameErrorStatus: false,
      genderErrorStatus: false,
      addressErrorStatus: false,
      roleErrorStatus: false,
      dateofbirthErrorStatus: false,
    };

    if (this.state.email.indexOf("@") === -1) {
      isError = true;
      errors.emailErrorStatus = true;
      errors.emailError = "Provide a valid email";
    }

    if (this.state.firstname.length <= 3) {
      console.info("VALUE", `this is the firstname ${this.state.firstname}`);
      isError = true;
      errors.firstnameErrorStatus = true;
      errors.firstnameError = "over 3 characters required";
    }

    if (this.state.lastname.length <= 3) {
      isError = true;
      errors.lastnameErrorStatus = true;
      errors.lastnameError = "over 3 characters required";
    }
    if (this.state.gender.length <= 3) {
      isError = true;
      errors.genderErrorStatus = true;
      errors.genderError = "over 3 characters required";
    }
    if (this.state.address.length <= 3) {
      isError = true;
      errors.addressErrorStatus = true;
      errors.addressError = "over 3 characters required";
    }
    if (this.state.role.length <= 3) {
      isError = true;
      errors.roleErrorStatus = true;
      errors.roleError = "over 3 characters required";
    }
    if (this.state.dateofbirth.length <= 3) {
      isError = true;
      errors.dateofbirthErrorStatus = true;
      errors.dateofbirthError = "over 3 characters required";
    }

    this.setState({
      ...this.state,
      ...errors,
    });
    return isError;
  };

  onSubmit = async (e) => {
    e.preventDefault();
    const error = this.validation();
    if (!error) {
      const data = {
        email: this.state.email,
        firstname: this.state.firstname,
        lastname: this.state.lastname,
        gender: this.state.gender,
        address: this.state.address,
        role: this.state.role,
        dateofbirth: this.state.dateofbirth,
      };
      await this.props.registrationAction(data);
    }
  };

  change = (e) => {
    e.preventDefault();
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    return (
      <div>
        <NavigationBar />
        {/* <p style={{ display: this.props.registrationState.loading }}>
          <Spinner animation="border" size="lg" />
        </p> */}
        <div id="registration">
          <div id="form">
            <form onSubmit={(e) => this.onSubmit(e)}>
              <h1>Registration</h1> <br />
              <label htmlFor="">Email</label>
              <br />
              <input
                type="email"
                name="email"
                aria-label="email"
                placeholder="email"
                onChange={(e) => this.change(e)}
              />{" "}
              <br />
              <Badge variant="danger">{this.state.emailError}</Badge>
              <br />
              <label htmlFor="">First name</label> <br />
              <input
                type="text"
                name="firstname"
                aria-label="firstname"
                placeholder="firstname"
                onChange={(e) => this.change(e)}
              />
              <br />
              <Badge variant="danger">{this.state.firstnameError}</Badge>
              <br />
              <label htmlFor="">Last name</label> <br />
              <input
                type="text"
                name="lastname"
                aria-label="lastname"
                placeholder="lastname"
                onChange={(e) => this.change(e)}
              />
              <br />
              <Badge variant="danger">{this.state.lastnameError}</Badge>
              <br />
              <label htmlFor="">Gender</label> <br />
              <select
                type="text"
                name="gender"
                aria-label="gender"
                placeholder="Gender"
                onChange={(e) => this.change(e)}
              >
                <option role="male">MALE</option>
                <option>FEMALE</option>
                <option>UNDEFINED</option>
              </select>
              <br />
              <Badge variant="danger">{this.state.genderError}</Badge>
              <br />
              <label htmlFor="">Address</label> <br />
              <input
                type="text"
                name="address"
                aria-label="address"
                placeholder="address"
                onChange={(e) => this.change(e)}
              />
              <br />
              <Badge variant="danger">{this.state.addressError}</Badge>
              <br />
              <label htmlFor="">Role</label> <br />
              <input
                type="text"
                name="role"
                aria-label="role"
                placeholder="role"
                onChange={(e) => this.change(e)}
              />
              <br />
              <Badge variant="danger">{this.state.roleError}</Badge>
              <br />
              <label htmlFor="">Date of Birth</label> <br />
              <input
                type="date"
                name="dateofbirth"
                aria-label="dateofbirth"
                placeholder="Date of birth"
                onChange={(e) => this.change(e)}
              />
              <br />
              <Badge variant="danger">{this.state.dateofbirthError}</Badge>
              <br />
              <input type="submit" value="SIGN UP" aria-label="signup" />
            </form>
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    registrationState: state.registration,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    registrationAction: (data) => dispatch(registrationAction(data)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Registration);
