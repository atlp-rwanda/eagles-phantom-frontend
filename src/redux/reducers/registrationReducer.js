import { REGISTER, REGISTER_SUCCESS, REGISTER_FAIL} from "../types/registrationTypes";

const initialState = {
    loading: 'none',
    data: [],
    error: ''
};

const registration = (state = initialState, action) => {
  switch (action.type) {
    case REGISTER:
      return {
        ...state,
        loading: 'block'
      };
    case REGISTER_SUCCESS:
      return {
        loading: 'none',
        data: action.payload,
        error: ''
      };
    case REGISTER_FAIL:
      return {
        loading: 'none',
        data: [],
        error: action.payload
      };

    default:
      return state;
  }
};

export default registration;