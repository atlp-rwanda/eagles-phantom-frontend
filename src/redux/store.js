import { createStore, compose, applyMiddleware, combineReducers } from "redux";
import thunk from "redux-thunk";
import registrationReducer from "../redux/reducers/registrationReducer";

const rootReducer = combineReducers({
  register: registrationReducer,
});

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(thunk))
);
