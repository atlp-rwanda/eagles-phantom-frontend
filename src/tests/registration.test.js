import React from "react";
import { Provider } from "react-redux";
import registrationReducer from '../redux/reducers/registrationReducer';
import { render, cleanup ,waitFor} from "@testing-library/react";
import userEvent from '@testing-library/user-event';
import Registration from '../components/registration/Registration';
import store from '../redux/store';
const onSubmit = jest.fn();

const RegisterComponent = () => {
    return render(
      <Provider store={store}>
        <Registration onSubmit={onSubmit} />
      </Provider>
    );
  };

describe('testing the reducer', () => {
    afterEach(cleanup);
    it('should return the initial state', () => {
        expect(registrationReducer(undefined, {})).toEqual(
            {
                loading: 'none',
                data: [],
                error: ''
            }
        )
    });
    it("should render the register component", () => {
        const { asFragment } = RegisterComponent();
        expect(asFragment(<Registration />)).toMatchSnapshot();
      });

      it('should change state/value of the different input field', () => {
        const { getByLabelText } = RegisterComponent();
        const email = getByLabelText('email');
        const fistname = getByLabelText('firstname');
        const lastname = getByLabelText('lastname');
        const address = getByLabelText('address');
        const role = getByLabelText('role');
    
        userEvent.type(email, 'example@email.com');
        userEvent.type(fistname, 'Allybaba');
        userEvent.type(lastname, 'Bomayee');
        userEvent.type(address, 'Kigali');
        userEvent.type(role, 'operator');
    
        expect(email.value).toBe('example@email.com');
        expect(fistname.value).toBe('Allybaba');
        expect(lastname.value).toBe('Bomayee');
        expect(address.value).toBe('Kigali');
        expect(role.value).toBe('operator');
      });

      it('should return error if email is not provided', async () => {
        const { getByLabelText, container, getByText, debug } = RegisterComponent();
        const email = getByLabelText('email');
        const form = container.querySelector('form');
    
        userEvent.type(email, 'example@email.com');
    
        form.dispatchEvent(new Event('submit'));
    
        waitFor(() => expect(getByText('Provide a password')).toBeTruthy());
      });

      it('should return error if email is already in the system', async () => {
        const { getByLabelText, container, getByText, debug } = RegisterComponent();
        const email = getByLabelText('email');
        const button = getByLabelText('signup');
    
        userEvent.type(email, 'email@gmail.com');
    
        button.click(button);
    
        waitFor(() =>
          expect(getByText('email is already in the system')).toBeTruthy()
        );
      });
})