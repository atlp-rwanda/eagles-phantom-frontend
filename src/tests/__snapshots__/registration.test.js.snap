// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the reducer should render the register component 1`] = `
<DocumentFragment>
  <div>
    <nav
      class="nav navbar navbar-expand-lg navbar-dark"
    >
      <a
        class="navbar-brand"
        href="#home"
      >
        Phantom
      </a>
      <button
        aria-controls="responsive-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="navbar-collapse collapse"
        id="responsive-navbar-nav"
      >
        <div
          class="justify-content-end navbar-nav"
        >
          <div
            class="nav_link nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="#"
              href="#"
              role="button"
            >
              Home
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="#"
              href="#"
              role="button"
            >
              About
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="#"
              href="#"
              role="button"
            >
              Contact
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link"
              data-rb-event-key="#"
              href="#"
              role="button"
            >
              Login
            </a>
          </div>
        </div>
      </div>
      <form
        class=""
      >
        <select
          class="form-control"
        >
          <option>
            English
          </option>
          <option>
            Francais
          </option>
          <option>
            Kinyarwanda
          </option>
        </select>
      </form>
    </nav>
    <div
      id="registration"
    >
      <div
        id="form"
      >
        <form>
          <h1>
            Registration
          </h1>
           
          <br />
          <label
            for=""
          >
            Email
          </label>
          <br />
          <input
            aria-label="email"
            name="email"
            placeholder="email"
            type="email"
          />
           
          <br />
          <span
            class="badge badge-danger"
          />
          <br />
          <label
            for=""
          >
            First name
          </label>
           
          <br />
          <input
            aria-label="firstname"
            name="firstname"
            placeholder="firstname"
            type="text"
          />
          <br />
          <span
            class="badge badge-danger"
          />
          <br />
          <label
            for=""
          >
            Last name
          </label>
           
          <br />
          <input
            aria-label="lastname"
            name="lastname"
            placeholder="lastname"
            type="text"
          />
          <br />
          <span
            class="badge badge-danger"
          />
          <br />
          <label
            for=""
          >
            Gender
          </label>
           
          <br />
          <select
            aria-label="gender"
            name="gender"
            placeholder="Gender"
            type="text"
          >
            <option
              role="male"
            >
              MALE
            </option>
            <option>
              FEMALE
            </option>
            <option>
              UNDEFINED
            </option>
          </select>
          <br />
          <span
            class="badge badge-danger"
          />
          <br />
          <label
            for=""
          >
            Address
          </label>
           
          <br />
          <input
            aria-label="address"
            name="address"
            placeholder="address"
            type="text"
          />
          <br />
          <span
            class="badge badge-danger"
          />
          <br />
          <label
            for=""
          >
            Role
          </label>
           
          <br />
          <input
            aria-label="role"
            name="role"
            placeholder="role"
            type="text"
          />
          <br />
          <span
            class="badge badge-danger"
          />
          <br />
          <label
            for=""
          >
            Date of Birth
          </label>
           
          <br />
          <input
            aria-label="dateofbirth"
            name="dateofbirth"
            placeholder="Date of birth"
            type="date"
          />
          <br />
          <span
            class="badge badge-danger"
          />
          <br />
          <input
            aria-label="signup"
            type="submit"
            value="SIGN UP"
          />
        </form>
      </div>
    </div>
    <div
      class="footer"
    >
      <div
        class="footer_info"
      >
        <div
          class="footer_quick_link"
        >
          <a
            class="footer_links"
            href="/about"
          >
            <b>
               About Us
            </b>
             
          </a>
          <br />
          <br />
          <br />
          <a
            class="footer_links"
            href="/"
          >
             
            <b>
              Services
            </b>
             
          </a>
          <br />
          <br />
          <br />
          <a
            class="footer_links"
            href="/"
          >
             
            <b>
              Terms of Use and Conditions
            </b>
             
          </a>
          <br />
          <br />
        </div>
        <div
          class="social-medias"
        >
          <h4>
            Follows us on:
          </h4>
          <br />
          <li>
            <a
              href="#"
            >
              <i
                class="fa fa-facebook"
              />
              <b>
                 Facebook
              </b>
            </a>
          </li>
           
          <li>
            <a
              href="#"
            >
              <i
                class="fa fa-instagram"
              />
              <b>
                 Instagram
              </b>
            </a>
          </li>
           
          <li>
            <a
              href="#"
            >
              <i
                class="fa fa-twitter"
              />
              <b>
                 Twitter
              </b>
            </a>
          </li>
           
        </div>
        <div
          class="address"
        >
          <h4>
            Contact Us On:
          </h4>
          <h3>
            <a
              href="#"
            >
              <i
                class="fa fa-phone"
              />
               +(250)788-620-444
            </a>
          </h3>
          <h3>
            <a
              href="#"
            >
              <i
                class="fa fa-envelope"
              />
               phantom@gmail.com
            </a>
          </h3>
        </div>
        <div
          class="social-medias-1"
        >
          <li>
            <a
              href="#"
            >
              <i
                class="fa fa-facebook"
              />
            </a>
          </li>
           
          <li>
            <a
              href="#"
            >
              <i
                class="fa fa-instagram"
              />
            </a>
          </li>
           
          <li>
            <a
              href="#"
            >
              <i
                class="fa fa-twitter"
              />
            </a>
          </li>
           
        </div>
      </div>
      <br />
      <br />
      <div
        class="copyright__footer"
      >
        <h4>
          ©2020 Phantom. All Rights Reserved.
        </h4>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
